### DISCUSSION TITLE
How to load and query an existing db from OPFS?

### DISCUSSION THREAD

**Discussion by lumosminima** [rhashimoto/wa-sqlite#221] [Q&A] [answered]
Hi 👋, I'm new to world of sqlite-wasm and OPFS and interested in adding some offline persistence for my PWA, previously I was using `@sqlite.org/sqlite-wasm` where I could read a db file from OPFS with it's default "opfs" vfs but performance was bad and it needed some special http headers so I looked into "opfs-sahpool", performance is much better but it only worked on a single tab so I discovered the #81 and copied SharedWorker.js in my app which worked great until I opened the PWA on chrome on android and discovered that it doesn't support SharedWorkers, I then tried to use service worker approach but for some reason it didn't work and was giving infinite 404 request errors on clientId but since "opfs-sahpool" was inspired by this lib, I decided to use it directly and currently trying to use OPFSCoopSyncVFS since it says on comparison page that it handles multiple connections (multiple tabs?) by default but I can't figure out how to use an existing db that is present in OPFS and open using `open_v2`. I get following error:
```js
sqliteworker.ts:74 NotFoundError: A requested file or directory could not be found at the time an operation was processed.
+page.svelte:20 Uncaught (in promise) Error: sqlite3_open_v2
    at check (sqlite-api.js?v=0d90b6dd:856:11)
    at Object.open_v2 (sqlite-api.js?v=0d90b6dd:495:9)
    at async Object.openDb (sqliteworker.ts?work…e&type=module:61:12)
```
This is my `sqliteworker.ts`:
```TS
import * as Comlink from 'comlink';
import { Factory, SQLITE_OPEN_READWRITE } from 'wa-sqlite';
import SQLiteESMFactory from 'wa-sqlite/dist/wa-sqlite.mjs';
import { OPFSCoopSyncVFS } from 'wa-sqlite/src/examples/OPFSCoopSyncVFS';
const log = console.log;
const error = console.error;

const sqlite = () => {
	let db: number | null = null;
	let sqlite3: SQLiteAPI | null = null;

	async function addStaticAssetToOPFS(url, fileName) {
		try {
			const response = await fetch(url);
			const blob = await response.blob();
			const fileHash = await computeHash(blob);

			if (await fileExistsWithHash(fileName, fileHash)) {
				console.log(`File ${fileName} already exists with the same hash.`);
				return;
			}

			await writeFileToOPFS(fileName, blob);
			console.log(`Added ${fileName} to OPFS`);
		} catch (error) {
			console.error(`Error adding ${fileName} to OPFS:`, error);
		}
	}

	async function computeHash(file) {
		const buffer = await file.arrayBuffer();
		const hashBuffer = await crypto.subtle.digest('SHA-256', buffer);
		const hashArray = Array.from(new Uint8Array(hashBuffer));
		return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
	}

	async function fileExistsWithHash(fileName, hash) {
		const root = await navigator.storage.getDirectory();
		try {
			const fileHandle = await root.getFileHandle(fileName);
			const file = await fileHandle.getFile();
			const existingHash = await computeHash(file);
			return existingHash === hash;
		} catch (error) {
			// File doesn't exist
			return false;
		}
	}

	async function writeFileToOPFS(fileName, fileOrBlob) {
		const root = await navigator.storage.getDirectory();
		const fileHandle = await root.getFileHandle(fileName, { create: true });
		const writable = await fileHandle.createWritable({ keepExistingData: false });
		await writable.write(fileOrBlob);
		await writable.close();
	}

	const preparation = async () => {
		const SQLiteEMSModule = await SQLiteESMFactory();
		const sqlite = Factory(SQLiteEMSModule);
		const vfs = await OPFSCoopSyncVFS.create('sq-vfs', SQLiteEMSModule);

		// vfs.jOpen('datatest.db');

		sqlite.vfs_register(vfs, true);

		return sqlite;
	};

	async function openDb() {
		if (!db) {
			sqlite3 = await preparation();

			db = await sqlite3.open_v2('datatest.db', SQLITE_OPEN_READWRITE, 'sq-vfs');
		}
	}

	async function closeDb() {
		if (db) {
			sqlite3?.close(db);
		}
	}

	return {
		addStaticAssetToOPFS,
		openDb,
		closeDb,
	};
};

export type SqliteType = ReturnType<typeof sqlite>;
const sqliteInstance = sqlite();
Comlink.expose(sqliteInstance);
```
And this is how I call:
```TS
		await sqlite.addStaticAssetToOPFS(
			`<db-blob-url>`,
			'datatest.db',
			true
		);
               await sqlite.openDb()
```
I tried to read the `file` demo in this repo but I don't quite understand it since it is using some API's like Streams API which I'm not familiar with, If there is currently no way to directly to above then some noob friendly example code on this would be much appreciated! ☺️

**Comment by rhashimoto** [comment]
I think I understand the issue. It is a peculiarity with OPFSCoopSyncVFS, which has to work some magic to provide synchronous VFS methods even though some of its operations - like opening a file - are asynchronous.

You can try one of two workarounds:

1. Add SQLITE_OPEN_CREATE in addition to SQLITE_OPEN_READWRITE to your open_v2() flags.
2. When you add your file "datatest.db" to OPFS, also add empty files "datatest.db-journal" and "datatest.db-wal".

If either of these workarounds fixes your problem then I'll attempt to explain why.

**Reply by lumosminima** [reply]
With first workaround I use: `await sqlite3.open_v2('datatest.db', SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, 'sq-vfs')` and get a new 0B datatest.db file and afterward get this error for actual db file even though I add to OPFS first and open db after:
```
Error adding datatest.db to OPFS: NoModificationAllowedError: Failed to execute 'createWritable' on 'FileSystemFileHandle': An attempt was made to modify an object where modifications are not allowed.
```
For second workaround, I can see db with correct size along with journal file that I manually add but after any exec queries I get following error: (except for test queries like `SELECT 'Hello, world!'`)
```
Uncaught (in promise) Error: unable to open database file
    at check (sqlite-api.js?v=15e7ac56:856:11)
    at sqlite-api.js?v=15e7ac56:708:13
    at async Factory.sqlite3.exec (sqlite-api.js?v=15e7ac56:419:22)
    at async Object.run (sqliteworker.ts?worker_file&type=module:134:7)
```
On further observation, I noticed that the `unable to open` error also sometimes happen in first workaround if db sometimes get successfully added to OPFS (_I check for db size using OPFS explorer extension_)

**Reply by rhashimoto** [reply]
I'll try to take a look later this week.

**Reply by rhashimoto** [reply, marked as answer]
Using SQLITE_OPEN_CREATE seems to work for me.

https://github.com/rhashimoto/opfs-import-test
https://github.com/rhashimoto/opfs-import-test/blob/master/worker.js

**Reply by lumosminima** [reply]
Thanks for the example 🙏, it was working with the `test.db` that is present but when I replaced it with my db in directory I was still getting error and on inspecting test.db I found that it was not WAL-enabled and my generated db was which is why I was getting `unable to open errors` I wasn't getting such error on `@sqlite.org/sqlite-wasm` because it's `importDb()` method for `opfs-sahpool` ["forces db out of WAL mode"](https://sqlite.org/wasm/doc/trunk/persistence.md#vfs-opfs-sahpool:~:text=written%20is%20returned.-,If%20the%20imported%20database%20is%20in%20WAL%20mode%20then%20it%20is%20forced%20out%20of%20WAL%20mode,-for%20historical%20reasons).

**Reply by rhashimoto** [reply]
Ah, I have never tried to import a WAL database so that never crossed my mind. I'm glad you figured it out.

Note that you should be able to use WAL if the first SQL you execute is:

```
PRAGMA locking_mode=EXCLUSIVE;
```

Then you can either work with the database in WAL mode or switch it to another journal_mode.

**Reply by sgbeal** [reply]
> Note that you should be able to use WAL if the first SQL you execute is:

In the mean time that works with that API but at the time importDb() was implemented it didn't, so importDb() is now constrained by the backwards-compatible behavior. Users can hypothetically now active exclusive locking and change the journal mode _back_ to WAL if they want to.
